fem/bilininteg_ext.cpp

```diff
@@ -0,0 +1,1581 @@
+               {
+                  gradXY[dy][dx][0] += gradX[dx][0] * wy;
+                  gradXY[dy][dx][1] += gradX[dx][1] * wDy;
+                  gradXY[dy][dx][2] += gradX[dx][2] * wy;
+               }
+            }
+         }
+         for (int dz = 0; dz < D1D; ++dz)
+         {
+            const double wz  = Bt(dz,qz);
+            const double wDz = Gt(dz,qz);
+            for (int dy = 0; dy < D1D; ++dy)
+            {
+               for (int dx = 0; dx < D1D; ++dx)
+               {
+                  y(dx,dy,dz,e) +=
+                     ((gradXY[dy][dx][0] * wz) +
+                      (gradXY[dy][dx][1] * wz) +
+                      (gradXY[dy][dx][2] * wDz));
+               }
+            }
+         }
+      }
+   });
+}
+
+typedef void (*fDiffusionMultAdd)(const int NE,
+                                  const double* B,
+                                  const double* G,
+                                  const double* Bt,
+                                  const double* Gt,
+                                  const double* oper,
+                                  const double* solIn,
+                                  double* solOut);
+
+static void PADiffusionMultAssembled(const int dim,
+                                     const int D1D,
+                                     const int Q1D,
+                                     const int NE,
+                                     const double* B,
+                                     const double* G,
+                                     const double* Bt,
+                                     const double* Gt,
+                                     const double* op,
+                                     const double* x,
+                                     double* y)
+{
+#ifdef MFEM_USE_OCCA
+   if (Device::UsingOcca())
+   {
+      MFEM_ASSERT(dim == 2, "");
+      OccaPADiffusionMultAdd2D(D1D, Q1D, NE, B, G, Bt, Gt, op, x, y);
+      return;
+   }
+#endif // MFEM_USE_OCCA
+   const int id = (dim<<8)|(D1D<<4)|(Q1D);
+   static std::unordered_map<int, fDiffusionMultAdd> call =
+   {
+      // 2D
+      {0x222,&PADiffusionMultAssembled2D<2,2>},
+      {0x233,&PADiffusionMultAssembled2D<3,3>},
+      {0x244,&PADiffusionMultAssembled2D<4,4>},
+      // 3D
+      {0x323,&PADiffusionMultAssembled3D<2,3>},
+   };
+   if (!call[id])
+   {
+      printf("\n%s:%d\nUnknown kernel with dim=%d, D1D=%d and Q1D=%d",
+             __FILE__, __LINE__, dim, D1D, Q1D);
+      mfem_error("PADiffusionMultAssembled kernel not instanciated");
+   }
+   call[id](NE, B, G, Bt, Gt, op, x, y);
+}
+
+// PA Diffusion MultAdd kernel
+void DiffusionIntegrator::MultAssembled(Vector &x, Vector &y)
+{
+   PADiffusionMultAssembled(dim, dofs1D, quad1D, ne,
+                            maps->B, maps->G, maps->Bt, maps->Gt,
+                            vec, x, y);
+}
+
+
+// PA Mass Assemble kernel
+void MassIntegrator::Assemble(const FiniteElementSpace &fes)
```

Comment:
Wasn't there an occa version of this?

Reply:
Good question -- @camierjs and @dmed256, didn't the Mass integrator also get ported?

Reply:
I just brought the diffusion kernel for `ex1` as an example on how it could  be done as a first try. The mass one can be brought for sure from the `occa-dev` branch.

Reply:
@dmed256, don't you want to bring in more *.okl kernels?